_posts/2016-08-31-continuous-delivery-databases.md

```diff
@@ -0,0 +1,57 @@
+
+Generell benötigt man dazu ein Schemamigrationstool, denn auch die Schemamigration muss man automatisieren.
+Dazu eignen sich Werkzeuge, wie
+
+* [Liquibase](http://www.liquibase.org/)
+* [Flyway](https://flywaydb.org/)
+* oder wenn man Oracle Datenbanken verwaltet [Orcas](https://github.com/opitzconsulting/orcas)
+
+Natürlich kann man mit Liquibase und Flyway auch Oracle Datenbanken verwalten. Orcas kann dann aber eine bessere Wahl sein.
+Insbesondere, wenn man die [unterschiedliche Philosophie](http://opitzconsulting.github.io/orcas/docs/de/liquibase/) von Orcas gegenüber Flyway und Liquibase vorzieht.
+Aber im Folgenden abstrahieren wir vom konkreten Produkt und zeigen wie man die Schemamigration in eine Continuous Delivery Pipeline
+einbauen kann.
+
+## Anforderungen an die Integration
+
+Hier einige Anforderungen und Randbedingungen, die für uns wichtig waren, als wir das Schemamanagement integriert haben.
+
+* Es werden die gleichen Mechanismen für alle Umgebungen (Entwicklung, Test, User Acceptance Test, Capacity Test, Pre-Production, Production) verwendet, um möglichst häufig die Migration zu testen, Inkonsistenzen durch unterschiedliches Vorgehen und Mehraufwand durch doppelte Pflege zu vermeiden.
+* Die Migration wird als einmaliger Schritt im Rahmen des Deployments einer Umgebung betrachtet
+* Es muss weitgehend möglich sein, die Migration von persistenten Umgebungen (Datenbank bleibt beim Neudeployment der Umgebung bestehen) vorher zu testen. Dies betrifft also insbesondere PreProduction und Production
+* Die Migration muss automatisiert durch Skripte erfolgen, damit sich durch manuelle Eingriffe keine Fehler einschleichen und der Aufwand gering bleibt.
+* Wir müssen auch "Basisdaten" verwalten können. Dies umfasst z.B. Referenztabellen (z.B. Ländercodes).
+
+## Kompromisse
+
+Häufig müssen bei Qualitätskritieren Kompromisse eingeangen werden. Ein für uns sehr wichtiger Kompromiss ist der zwischen Qualität der Datenmigration und Dauer für bestimmte Schritte in der Delivery Pipeline.
+
+Da wir für alle Umgebungen ein gleichartiges Vorgehen nutzen, bedeutet dies für einige Umgebungen, insbesondere bei denen es auch eine ephemerale Datenbank gibt, dass das Erzeugen des Schemas länger dauert als wenn man die Datenbank durch ein einziges Skript auf den aktuell gültigen Zustand bringt und nicht inkrementelle Änderungen anwendet (die ja mit der Zeit immer länger werden).
+Dies hat Auswirkungen z.B. auf die Dauer von E2E Tests. Um dies zu kompensieren, kann die Nutzung von mehr Ressourcen notwendig sein. Die Zusatzkosten sind aber unserer Erfahrung nach geringer als die Kosten aufgrund doppelter Pflege und
+Nacharbeiten aufgrund schlechter Qualität.  
+
+## Lösungsmöglichkeit
+
+Wir verwenden eine komplett auf Docker basierende Continuous Delivery Pipeline. Dies bedeutet, dass wir ab einem bestimmten Schritt nicht
+mehr mit den jar / war Artefakten arbeiten, sondern dirket mit den erzeugten Docker Images.
```

Comment:
Ich würde `JAR- oder WAR-Artefakten` schrieben.

Typo in `dirket`, gemeint ist wohl `direkt`.